wpilib/tests/test_ultrasonic.py

```diff
@@ -1,27 +1,155 @@
+    assert ultrasonic.getRangeMM() == pytest.approx(206.5, 0.01)
+
+
+def test_ultrasonic_pidsourcetype_allowed(ultrasonic, wpilib, hal_data):
+    ultrasonic.setPIDSourceType(wpilib.interfaces.PIDSource.PIDSourceType.kDisplacement)
+    assert ultrasonic.getPIDSourceType() == wpilib.interfaces.PIDSource.PIDSourceType.kDisplacement
+
+
+def test_ultrasonic_pidsourcetype_notallowed(ultrasonic, wpilib):
+    with pytest.raises(ValueError) as ex:
+        ultrasonic.setPIDSourceType(wpilib.interfaces.PIDSource.PIDSourceType.kRate)
+
+
+@pytest.mark.parametrize("unit_name, expected_output", [
+    ('kInches', 8.13),
+    ('kMillimeters', 206.5),
+    ])
+def test_ultrasonic_pidget(ultrasonic, wpilib, hal_data, unit_name, expected_output):
+    ultrasonic.ping()
+    hal_data['counter'][0]['count'] = 2
+    hal_data['counter'][0]['period'] = 0.0012
+    unit = getattr(wpilib.Ultrasonic.Unit, unit_name)
+    ultrasonic.setDistanceUnits(unit)
+    assert ultrasonic.getDistanceUnits() == unit
+    assert ultrasonic.pidGet() == pytest.approx(expected_output, 0.01)
+
+
+def test_ultrasonic_pidget_invalid(ultrasonic):
+    with pytest.raises(ValueError) as ex:
+        ultrasonic.setDistanceUnits(-1)
+
+
+def test_ultrasonic_inittable_nulltable(ultrasonic):
+    ultrasonic.initTable(None)
+
+    assert ultrasonic.valueEntry is None
+
+
+def test_ultrasonic_inittable(ultrasonic):
+    table = MagicMock()
+    ultrasonic.updateTable = MagicMock()
     
-    u.free()
-    
\ No newline at end of file
+    ultrasonic.initTable(table)
+
+    table.getEntry.assert_has_calls([call("Value")])
+    assert ultrasonic.updateTable.called
+
+
+def test_ultrasonic_updateTable_nullentry(ultrasonic):
+    ultrasonic.getPressureSwitchValue = MagicMock()
+    ultrasonic.value = MagicMock()
+    ultrasonic.updateTable()
+
+    assert not ultrasonic.getPressureSwitchValue.called
+    assert not ultrasonic.value.called
+
+
+def test_ultrasonic_updateTable(ultrasonic, hal_data):
+    ultrasonic.ping()
+    hal_data['counter'][0]['count'] = 2
+    hal_data['counter'][0]['period'] = 0.0012
+
+    ultrasonic.valueEntry = MagicMock()
+
+    ultrasonic.updateTable()
+
+    ultrasonic.valueEntry.setDouble.assert_called_with(pytest.approx(8.13, 0.01))
+
+
+def test_startlivewindowmode(ultrasonic):
+    ultrasonic.valueEntry = MagicMock()
+    ultrasonic.startLiveWindowMode()
+
+
+def test_startlivewindowmode_null_entry(ultrasonic):
+    ultrasonic.valueEntry = None
+
+    ultrasonic.startLiveWindowMode()
+
+
+def test_stoplivewindowmode(ultrasonic):
+    ultrasonic.valueEntry = MagicMock()
+
+    ultrasonic.startLiveWindowMode()
+
+    assert not hasattr(ultrasonic, 'valueListener')
+
+    ultrasonic.stopLiveWindowMode()
+
+    assert not hasattr(ultrasonic, 'valueListener')
+
+
+def test_stoplivewindowmode_null_entry(ultrasonic):
+    ultrasonic.valueEntry = None
+
+    ultrasonic.startLiveWindowMode()
+
+    ultrasonic.stopLiveWindowMode()
```

Comment:
This looks like it belongs in tests for `LiveWindowSendable` instead.

Reply:
that assumes Ultrasonic subclasses LiveWindowSendable, which is an implementation detail that might change in the future

Reply:
I think it belongs here.

Reply:
I'm still slightly sceptical, because there shouldn't be a scenario where the `startLiveWindowMode()` method on an Ultrasonic object should be called before `initTable()`.